packages/workflow/src/workflow.ts

```diff
@@ -526,3 +526,66 @@ export function uuid4(): string {
     4
   )}-${ho(view.getUint32(10), 8)}${ho(view.getUint16(14), 4)}`;
 }
+
+/**
+ * Patch or upgrade workflow code by checkin or stating that this workflow has a certain patch.
+ *
+ * TODO: See docs page for info
+ *
+ * If the workflow is replaying an existing history, then this function returns true if that
+ * history was produced by a worker which also had a `patched` call with the same `patchId`.
+ * If the history was produced by a worker *without* such a call, then it will return false.
+ *
+ * If the workflow is not currently replaying, then this call *always* returns true.
+ *
+ * Your workflow code should run the "new" code if this returns true, if it returns false, you
+ * should run the "old" code. By doing this, you can maintain determinism.
+ *
+ * @param patchId An identifier that should be unique to this patch. It is OK to use multiple
+ * calls with the same ID, which means all such calls will always return the same value.
+ */
+export function patched(patchId: string): boolean {
+  return patchInternal(patchId, false);
+}
+
+/**
+ * Indicate that a patch is being phased out.
+ *
+ * // TODO: See docs page, how to run queries, etc.
+ *
+ * Workflows with this call may be deployed alongside workflows with a {@link patched} call, but
+ * they must *not* be deployed while any workers still exist running old code without a
+ * {@link patched} call, or any runs with histories produced by such workers exist. If either kind
+ * of worker encounters a history produced by the other, their behavior is undefined.
+ *
+ * Once all live workflow runs have been produced by workers with this call, you can deploy workers
+ * which are free of either kind of patch call for this ID. Workers with and without this call
+ * may coexist, as long as they are both running the "new" code.
+ *
+ * @param patchId An identifier that should be unique to this patch. It is OK to use multiple
+ * calls with the same ID, which means all such calls will always return the same value.
+ */
+export function deprecatePatch(patchId: string): void {
+  patchInternal(patchId, true);
+}
+
+function patchInternal(changeId: string, deprecated: boolean): boolean {
+  // Always send the command to core, it's smart enough to ignore it if it was pointless.
```

Comment:
I wouldn't always send the command to core, it's pointless.

Reply:
See other comment

Reply:
There are 2 cases I think we're doing wrong here:
1. We keep sending the same marker over and over again
2. We send the marker even if the workflow used the old code

I think we should change this to:

```ts
  if (state.info === undefined) {
    throw new IllegalStateError('Workflow info must be set when calling patch functions');
  }
  const { isReplaying } = state.info;
  const usePatch = !isReplaying || state.knownPresentChanges.has(patchId);
  if (usePatch && !state.sentPatches.has(patchId)) {
    state.commands.push({
      setPatchMarker: {
        patchId: patchId,
        deprecated: deprecated,
      },
    });
    state.sentPatches.add(patchId);
  }
  return usePatch;
```

Reply:
The thing is though it just doesn't matter. Core cleans that up, and like I mentioned there's only an incredibly small amount of overhead b/c we still have to call core the same amount of times. So at the end of the day I think this is more code for very little benefit. The way it is it's simple to test, idempotent, etc. I guess I'm just not following why we'd want to add the complexity when we don't need to?

Reply:
EX: The code you posted doesn't work in the `replay-no-marker-then-not-replay patched` test. It's just tricky to get them all right. I've updated it to a version that just doesn't issue them more than once which I think is 90% of the benefit without the tricky logic.

It's also just easy to test -- first time it always gets sent, no need to think too hard